mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,145 +166,301 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
+            Value shadowPtr = shadowDesc.alignedPtr(builder, loc);
+
+            if (options.zeroOut) {
+                Value bufferSizeBytes =
+                    computeMemRefSizeInBytes(memRefType, shadowDesc, builder, loc);
+                Value zero = builder.create<LLVM::ConstantOp>(loc, builder.getI8Type(), 0);
+                builder.create<LLVM::MemsetOp>(loc, shadowPtr, zero, bufferSizeBytes,
+                                               /*isVolatile=*/false);
+            }
             callArgs.push_back(shadowPtr);
         }
+        else {
+            callArgs.push_back(enzymeConst);
+            callArgs.push_back(desc.alignedPtr(builder, loc));
+        }
 
-        // Results of callee and their shadow
-        SmallVector<Value> calleeArgs(op.getArgs().begin(), op.getArgs().end());
-        ValueRange results = rewriter.create<func::CallOp>(loc, callee, calleeArgs).getResults();
-
-        // Lower callee results to to llvm
-        SmallVector<Type> resCalleeTypes(callee.getResultTypes().begin(),
-                                         callee.getResultTypes().end());
-        for (auto resTypeIt = resCalleeTypes.begin(); resTypeIt < resCalleeTypes.end();
-             resTypeIt++) {
-            Type llvmResType = llvmTypeConverter.convertType(*resTypeIt);
-            if (!llvmResType)
-                emitError(loc, "Could not convert argmap result to LLVM type: ") << *resTypeIt;
-            *resTypeIt = llvmResType;
+        // Offsets, sizes, and strides
+        callArgs.push_back(desc.offset(builder, loc));
+        for (int64_t dim = 0; dim < memRefType.getRank(); dim++) {
+            callArgs.push_back(desc.size(builder, loc, dim));
         }
+        for (int64_t dim = 0; dim < memRefType.getRank(); dim++) {
+            callArgs.push_back(desc.stride(builder, loc, dim));
+        }
+    }
 
-        ValueRange llvmresults =
-            rewriter.create<UnrealizedConversionCastOp>(loc, resCalleeTypes, results).getResults();
-
-        // Store the results
-        for (auto [result, llvmresult, llvmshadowresult] :
-             llvm::zip(results, llvmresults, adaptor.getQuantumJacobian())) {
-            auto newArg = rewriter.create<LLVM::AllocaOp>(
-                loc, LLVM::LLVMPointerType::get(llvmresult.getType()), c1);
-            rewriter.create<LLVM::StoreOp>(loc, llvmresult, newArg);
-            callArgs.push_back(newArg);
-
-            // Add shadow for memref
-            Value shadowPtr = rewriter.create<LLVM::AllocaOp>(
-                loc, LLVM::LLVMPointerType::get(llvmshadowresult.getType()), c1);
-            rewriter.create<LLVM::StoreOp>(loc, llvmshadowresult, shadowPtr);
-            callArgs.push_back(shadowPtr);
+    Value computeMemRefSizeInBytes(MemRefType type, MemRefDescriptor descriptor, OpBuilder &builder,
+                                   Location loc) const
+    {
+        // element_size * (offset + sizes[0] * strides[0])
```

Comment:
Is this only for the first axis? Are we guaranteeing that memref is only of rank 1?

Reply:
Actually, it is for the last axis, the axis order start at the last one, don't they?

Reply:
This is for a general MemRef. Consider a `memref<MxNxK>`: the sizes will be `[M, N, K]` and the strides will be `[N*K, K, 1]`. The overall number of elements, `M * N * K`, can be obtained by multiplying the `sizes[0] * strides[0]`.

We can update the comment here to add more detail if it's confusing.

Reply:
Yes, a comment would be nice. Thanks @pengmai!